serviceregistration/kubernetes/service_registration.go

```diff
@@ -0,0 +1,238 @@
+				Path:      pathToLabels + labelActive,
+				Value:     toString(r.initialState.IsActive),
+			},
+			{
+				Operation: client.Replace,
+				Path:      pathToLabels + labelSealed,
+				Value:     toString(r.initialState.IsSealed),
+			},
+			{
+				Operation: client.Replace,
+				Path:      pathToLabels + labelPerfStandby,
+				Value:     toString(r.initialState.IsPerformanceStandby),
+			},
+			{
+				Operation: client.Replace,
+				Path:      pathToLabels + labelInitialized,
+				Value:     toString(r.initialState.IsInitialized),
+			},
+		}
+		if err := c.PatchPod(r.namespace, r.podName, patches...); err != nil {
+			return err
+		}
+	}
+
+	// Run a service that retries errored-out notifications if they occur.
+	go r.retryHandler.Run(shutdownCh, wait)
+
+	// Run a background goroutine to leave labels in the final state we'd like
+	// when Vault shuts down.
+	go r.onShutdown(shutdownCh, wait)
+
+	return nil
+}
+
+func (r *serviceRegistration) NotifyActiveStateChange(isActive bool) error {
+	return r.notifyOrRetry(&client.Patch{
+		Operation: client.Replace,
+		Path:      pathToLabels + labelActive,
+		Value:     toString(isActive),
+	})
+}
+
+func (r *serviceRegistration) NotifySealedStateChange(isSealed bool) error {
+	return r.notifyOrRetry(&client.Patch{
+		Operation: client.Replace,
+		Path:      pathToLabels + labelSealed,
+		Value:     toString(isSealed),
+	})
+}
+
+func (r *serviceRegistration) NotifyPerformanceStandbyStateChange(isStandby bool) error {
+	return r.notifyOrRetry(&client.Patch{
+		Operation: client.Replace,
+		Path:      pathToLabels + labelPerfStandby,
+		Value:     toString(isStandby),
+	})
+}
+
+func (r *serviceRegistration) NotifyInitializedStateChange(isInitialized bool) error {
+	return r.notifyOrRetry(&client.Patch{
+		Operation: client.Replace,
+		Path:      pathToLabels + labelInitialized,
+		Value:     toString(isInitialized),
+	})
+}
+
+func (r *serviceRegistration) onShutdown(shutdownCh <-chan struct{}, wait *sync.WaitGroup) {
+	// Make sure Vault will give us time to finish up here.
+	wait.Add(1)
+	defer wait.Done()
+
+	// Start running this when we receive a shutdown.
+	<-shutdownCh
+
+	// Label the pod with the values we want to leave behind after shutdown.
+	patches := []*client.Patch{
+		{
+			Operation: client.Replace,
+			Path:      pathToLabels + labelActive,
+			Value:     toString(false),
+		},
+		{
+			Operation: client.Replace,
+			Path:      pathToLabels + labelSealed,
+			Value:     toString(true),
+		},
+		{
+			Operation: client.Replace,
+			Path:      pathToLabels + labelPerfStandby,
+			Value:     toString(false),
+		},
+		{
+			Operation: client.Replace,
+			Path:      pathToLabels + labelInitialized,
+			Value:     toString(false),
+		},
+	}
+	if err := r.client.PatchPod(r.namespace, r.podName, patches...); err != nil {
+		if r.logger.IsError() {
+			r.logger.Error(fmt.Sprintf("unable to set final status on pod name %q in namespace %q on shutdown: %s", r.podName, r.namespace, err))
```

Comment:
Nit: I wonder if we actually need this final status.  Once the pod is destroyed due to Vault being unready, it will no longer exist and isn't even queryable.

Reply:
Ah! So we could simply not have that at all. That makes sense. Yes, I will delete the entire `onShutdown` method, the simpler the better.

Reply:
Stripped in https://github.com/hashicorp/vault/pull/8249/commits/708306b034b85b280fa408b87881452e5f39cc63